Detectors/EMCAL/workflow/src/EventBuilderSpec.cxx

```diff
@@ -0,0 +1,140 @@
+    auto trgrecorddata = ctx.inputs().getByPos(posTriggerRecords, islot);
+    auto subspecification = framework::DataRefUtils::getHeader<header::DataHeader*>(trgrecorddata)->subSpecification;
+    // Discard message if it is a deadbeaf message (empty timeframe)
+    if (subspecification == 0xDEADBEEF) {
+      continue;
+    }
+    triggers[subspecification] = ctx.inputs().get<gsl::span<o2::emcal::TriggerRecord>>(trgrecorddata);
+  }
+
+  for (auto interaction : connectRangesFromSubtimeframes(triggers)) {
+    eventCells.clear();
+    for (auto [subspec, triggerrecord] : interaction.mRangesSubtimeframe) {
+      LOG(DEBUG) << interaction.mInteractionRecord.bc << " / " << interaction.mInteractionRecord.orbit << ": Receiving " << triggerrecord.getEntries() << " digits for subspec " << subspec;
+      auto cellcont = cells.find(subspec);
+      if (cellcont != cells.end()) {
+        for (auto& cell : gsl::span<const Cell>(cellcont->second.data() + triggerrecord.getFirstEntry(), triggerrecord.getEntries())) {
+          eventCells.emplace_back(cell);
+        }
+      } else {
+        LOG(ERROR) << "No cell container found for subspec " << subspec;
+      }
+    }
+    int start = outputCells.size();
+    if (eventCells.size()) {
+      std::sort(eventCells.begin(), eventCells.end(), [](const Cell& lhs, const Cell& rhs) { return lhs.getTower() < rhs.getTower(); });
+      std::copy(eventCells.begin(), eventCells.end(), std::back_inserter(outputCells));
+    }
+    outputTriggers.emplace_back(interaction.mInteractionRecord, interaction.mTriggerType, start, eventCells.size());
+  }
+
+  ctx.outputs().snapshot(framework::Output{dataorigin, "CELLS", 0, framework::Lifetime::Timeframe}, outputCells);
+  ctx.outputs().snapshot(framework::Output{dataorigin, "CELLSTRGR", 0, framework::Lifetime::Timeframe}, outputTriggers);
+}
+
+std::set<EventBuilderSpec::RangeCollection> EventBuilderSpec::connectRangesFromSubtimeframes(const std::unordered_map<o2::header::DataHeader::SubSpecificationType, gsl::span<const TriggerRecord>>& triggerrecords) const
+{
+  std::set<RangeCollection> events;
+
+  // Search interaction records from all subevents
+  std::set<o2::InteractionRecord> allInteractions;
+  for (auto& [subspecification, trgrec] : triggerrecords) {
+    for (auto rec : trgrec) {
+      auto eventIR = rec.getBCData();
+      if (allInteractions.find(eventIR) == allInteractions.end())
+        allInteractions.insert(eventIR);
+    }
+  }
+
+  // iterate over all subevents for all bunch crossings
+  for (auto collision : allInteractions) {
+    RangeCollection nextevent;
+    nextevent.mInteractionRecord = collision;
+    bool first = true,
+         hasSubevent = false;
+    for (auto [subspecification, records] : triggerrecords) {
+      auto found = std::find_if(records.begin(), records.end(), [&collision](const o2::emcal::TriggerRecord& rec) { return rec.getBCData() == collision; });
```

Comment:
Not sure if the overhead is significant, but instead of using find for every int.record, you could use the fact that the data in each of triggerrecords span is ordered in int.record, i.e. something like this:
````
std::vector<int> cursor(triggerrecords.size());
for (auto collision : allInteractions) {
  RangeCollection nextevent;
  nextevent.mInteractionRecord = collision;
  bool first = true, hasSubevent = false;
  int count = 0;
  for (auto [subspecification, records] : triggerrecords) {
     if (cursor[count] >= records.size()) {
       count++;
       continue;
     }
     while (cursor[count]<records.size() && records[cursor[count]].getBCData() < collision) {
       cursor[count]++;
     }
     auto slotID  = cursor[count++];
     if (slotID>=records.size()) { // not found !
       continue;
     }
     .... use data of the trigger  slotID
  }
````

Reply:
@shahor02 I do the connection of the ranges now in a different way: When I iterate over trigger records from all subtimeframes I cache the index in the gsl span for the subspec range. The first find call is still kept, however as I am using a std::unordered_map for this the overhead should be small.